stablehlo/dialect/TypeInference.cpp

```diff
@@ -2069,6 +2172,280 @@ LogicalResult verifyCollectivePermuteOp(
+  if (failed(windowStridesOrErr)) return failure();
+  auto lhsDilationOrErr =
+      hlo::convert1DAttribute(lhsDilation, location, "lhs_dilation");
+  if (failed(lhsDilationOrErr)) return failure();
+  auto rhsDilationOrErr =
+      hlo::convert1DAttribute(rhsDilation, location, "rhs_dilation");
+  if (failed(rhsDilationOrErr)) return failure();
+  auto windowReversalOrErr = hlo::convertWindowReversalAttribute(
+      windowReversal, location, "window_reversal");
+  if (failed(windowReversalOrErr)) return failure();
+
+  auto windowOrErr = hlo::verifyWindowAttributesAndInferWindowDimensions(
+      windowDimensions, *windowStridesOrErr, *paddingOrErr, *lhsDilationOrErr,
+      *rhsDilationOrErr, *windowReversalOrErr, location);
+  if (failed(windowOrErr)) return failure();
+
+  // P3.
+  if (failed(verifyPrecisionConfig(location, precisionConfig)))
+    return failure();
+
+  // P5.
+  SmallVector<int64_t> outputDimensions(lhsType.getShape().size(),
+                                        ShapedType::kDynamic);
+
+  // Infer the output spatial dimensions.
+  auto numSpatialDims = inputSpatialDimensions.size();
+  SmallVector<int64_t> inputSpatialDimVals(numSpatialDims);
+  for (int64_t i = 0; i < static_cast<int64_t>(numSpatialDims); ++i)
+    inputSpatialDimVals[i] = lhsType.getShape()[inputSpatialDimensions[i]];
+
+  auto windowOutputShape =
+      hlo::inferWindowOutputShape(inputSpatialDimVals, *windowOrErr);
+
+  for (int64_t i = 0; i < static_cast<int64_t>(windowOrErr->size()); ++i)
+    outputDimensions[outputSpatialDimensions[i]] = windowOutputShape[i];
+
+  // Infer the output-batch-dimension and output-feature-dimension.
+  const int64_t inputBatch = lhsType.getShape()[inputBatchDimension];
+  const int64_t kernelOutputFeatures =
+      rhsType.getShape()[kernelOutputFeatureDimension];
+
+  outputDimensions[outputBatchDimension] = hlo::isDynamicDimSize(inputBatch)
+                                               ? ShapedType::kDynamic
+                                               : inputBatch / batchGroupCount;
+  outputDimensions[outputFeatureDimension] = kernelOutputFeatures;
+
+  auto resultType = result.getType().dyn_cast<RankedTensorType>();
+  if (!resultType) return success();
+
+  if (failed(verifyCompatibleShape(outputDimensions, resultType.getShape())))
+    return emitOptionalError(
+        location, "inferred shape '", dimSizesToString(outputDimensions), "' ",
+        "is incompatible with return type of operation ", resultType, "");
+
+  return success();
+}
+
+LogicalResult verifyDotOp(Optional<Location> location, Value lhs, Value rhs,
+                          Optional<ArrayAttr> /*precisionConfig*/,
+                          Value result) {
+  auto lhsType = lhs.getType().dyn_cast<RankedTensorType>();
+  auto rhsType = rhs.getType().dyn_cast<RankedTensorType>();
+  auto resultType = result.getType().dyn_cast<RankedTensorType>();
+  if (!lhsType || !rhsType || !resultType) return success();
```

Comment:
Your `verifyConvolutionop()` inspire me that this `if (!resultType) return success() ` should show up after all the verifications are done for the inputs/attrs, right? I mean to move it below the `inferDotShape()` part.